File: config/manager_config/go/selector.go
// Package manager resolves a component's environment configuration from one variable, and
// validates it before returning a value.
//
// This mirrors config/manager_config/rust. Where the two disagree, one of them is a bug: the
// same SERVICERADAR_ENV must yield the same identity, the same source, and the same errors in
// every language, or a deployment behaves differently depending on which service reads it.
package manager

import (
	"fmt"
	"os"
	"strings"
)

// EnvVar is the only environment variable a component reads to determine its configuration.
const EnvVar = "SERVICERADAR_ENV"

const onprem = "onprem"

// Kinds that do not accept an instance identifier. onprem requires one.
var singleInstanceKinds = []string{"localhost", "ci", "saas", "demo"}

// Identity is the environment, as <kind>[":" <instance>].
type Identity struct {
	Kind     string
	Instance string // empty when the kind is single-instance
}

func (i Identity) String() string {
	if i.Instance == "" {
		return i.Kind
	}
	return i.Kind + ":" + i.Instance
}

type SelectorErrorKind int

const (
	Absent SelectorErrorKind = iota
	UnknownKind
	InstanceRequired
	InstanceNotAccepted
)

type SelectorError struct {
	Kind  SelectorErrorKind
	Value string
}

func (e *SelectorError) Error() string {
	switch e.Kind {
	case Absent:
		return unsetMessage()
	case UnknownKind:
		return fmt.Sprintf("%s=%q names no environment kind. Valid: %s, %s:<instance>.",
			EnvVar, e.Value, strings.Join(singleInstanceKinds, ", "), onprem)
	case InstanceRequired:
		return fmt.Sprintf("%s=%q requires an instance identifier, as %s:<instance>.",
			EnvVar, e.Value, e.Value)
	case InstanceNotAccepted:
		return fmt.Sprintf("%s kind %q does not accept an instance identifier.", EnvVar, e.Value)
	}
	return "unknown selector error"
}

// unsetMessage is the one failure a reader may be meeting for the first time, possibly at 3am,
// in a crash loop with no other output. It says what is wrong, why nothing can proceed, exactly
// what to set, and how to set it on each platform -- because the reader's next action is editing
// a manifest, not reading source.
func unsetMessage() string {
	return fmt.Sprintf(`
==============================================================================
SERVICERADAR CANNOT START: %[1]s is not set.
==============================================================================

This one environment variable declares WHICH ServiceRadar environment this
process is running in. Everything else is derived from it: the database, the
message bus, the TLS posture, and which provider resolves secrets. Nothing can
be loaded until it is set.

There is deliberately NO DEFAULT. A guessed environment is a guessed database,
and guessing wrong is silent -- the process would start and connect somewhere
nobody chose.

Set %[1]s to exactly one of:

  %[2]s
  %[3]s:<instance>     (on-prem is multi-instance; name the deployment)

How to set it:

  Kubernetes   env:
                 - name: %[1]s
                   value: saas
  Docker       docker run -e %[1]s=saas ...
  Compose      environment:
                 %[1]s: saas
  CI           export %[1]s=ci
  Local dev    export %[1]s=localhost
==============================================================================`,
		EnvVar, strings.Join(singleInstanceKinds, "\n  "), onprem)
}

// ParseIdentity reads the identity from the variable's value.
//
// Takes the value rather than the environment so it stays a total function of its input and is
// testable without mutating process state.
func ParseIdentity(value string, present bool) (Identity, error) {
	// Empty is unset, not a choice: a shell exporting SERVICERADAR_ENV= has selected nothing,
	// and this repository's build tooling pins several variables to "" deliberately.
	trimmed := strings.TrimSpace(value)
	if !present || trimmed == "" {
		return Identity{}, &SelectorError{Kind: Absent}
	}

	kind, instance, hasInstance := strings.Cut(trimmed, ":")

	if kind == onprem {
		if !hasInstance || instance == "" {
			return Identity{}, &SelectorError{Kind: InstanceRequired, Value: onprem}
		}
		return Identity{Kind: kind, Instance: instance}, nil
	}
	if !contains(singleInstanceKinds, kind) {
		return Identity{}, &SelectorError{Kind: UnknownKind, Value: trimmed}
	}
	if hasInstance {
		return Identity{}, &SelectorError{Kind: InstanceNotAccepted, Value: kind}
	}
	return Identity{Kind: kind}, nil
}

// IdentityFromEnv reads the one variable from the process environment.
func IdentityFromEnv() (Identity, error) {
	value, present := os.LookupEnv(EnvVar)
	return ParseIdentity(value, present)
}

func contains(set []string, v string) bool {
	for _, s := range set {
		if s == v {
			return true
		}
	}
	return false
}

File: config/manager_config/rust/src/selector.rs
//! Which environment this process is, read from the one variable that decides it.
//!
//! `SERVICERADAR_ENV` is the ONLY input. It is required and has no default: a component that
//! guessed an environment would guess a database, and the wrong guess is silent. Everything
//! else -- which instance to load, where it comes from, which secret provider resolves
//! credentials -- is derived from the identity this yields.

use std::fmt;

pub const ENV_VAR: &str = "SERVICERADAR_ENV";

/// The kinds that do not accept an instance identifier. `onprem` requires one.
const SINGLE_INSTANCE_KINDS: &[&str] = &["localhost", "ci", "saas", "demo"];
const ONPREM: &str = "onprem";

/// The environment identity, as `<kind>[":" <instance>]`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Identity {
    pub kind: String,
    pub instance: Option<String>,
}

impl fmt::Display for Identity {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match &self.instance {
            Some(i) => write!(f, "{}:{}", self.kind, i),
            None => write!(f, "{}", self.kind),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SelectorError {
    Absent,
    UnknownKind(String),
    InstanceRequired(String),
    InstanceNotAccepted(String),
}

impl fmt::Display for SelectorError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            // The one failure a reader may be meeting for the first time, possibly at 3am, in a
            // crash loop with no other output. It says what is wrong, why nothing can proceed,
            // exactly what to set, and how to set it on each platform -- because the reader's
            // next action is editing a manifest, not reading source.
            Self::Absent => write!(
                f,
                "\n\
                 ==============================================================================\n\
                 SERVICERADAR CANNOT START: {ENV_VAR} is not set.\n\
                 ==============================================================================\n\
                 \n\
                 This one environment variable declares WHICH ServiceRadar environment this\n\
                 process is running in. Everything else is derived from it: the database, the\n\
                 message bus, the TLS posture, and which provider resolves secrets. Nothing can\n\
                 be loaded until it is set.\n\
                 \n\
                 There is deliberately NO DEFAULT. A guessed environment is a guessed database,\n\
                 and guessing wrong is silent -- the process would start and connect somewhere\n\
                 nobody chose.\n\
                 \n\
                 Set {ENV_VAR} to exactly one of:\n\
                 \n\
                   {}\n\
                   {ONPREM}:<instance>     (on-prem is multi-instance; name the deployment)\n\
                 \n\
                 How to set it:\n\
                 \n\
                   Kubernetes   env:\n\
                                  - name: {ENV_VAR}\n\
                                    value: saas\n\
                   Docker       docker run -e {ENV_VAR}=saas ...\n\
                   Compose      environment:\n\
                                  {ENV_VAR}: saas\n\
                   CI           export {ENV_VAR}=ci\n\
                   Local dev    export {ENV_VAR}=localhost\n\
                 ==============================================================================",
                SINGLE_INSTANCE_KINDS.join("\n                   ")
            ),
            Self::UnknownKind(k) => write!(
                f,
                "{ENV_VAR}={k:?} names no environment kind. Valid: {}, {ONPREM}:<instance>.",
                SINGLE_INSTANCE_KINDS.join(", ")
            ),
            Self::InstanceRequired(k) => {
                write!(f, "{ENV_VAR}={k:?} requires an instance identifier, as {k}:<instance>.")
            }
            Self::InstanceNotAccepted(k) => {
                write!(f, "{ENV_VAR} kind {k:?} does not accept an instance identifier.")
            }
        }
    }
}

impl std::error::Error for SelectorError {}

/// Parses the identity from the variable's value.
///
/// Takes the value rather than reading the environment so this stays a total function of its
/// input and is testable without mutating process state.
pub fn resolve(value: Option<&str>) -> Result<Identity, SelectorError> {
    // Empty is unset, not a choice: a shell exporting `SERVICERADAR_ENV=` has selected nothing,
    // and this repository's build tooling pins several variables to "" deliberately.
    let value = value.map(str::trim).filter(|s| !s.is_empty());
    let Some(value) = value else {
        return Err(SelectorError::Absent);
    };

    let (kind, instance) = match value.split_once(':') {
        Some((k, i)) => (k, Some(i.to_string())),
        None => (value, None),
    };

    if kind == ONPREM {
        return match instance {
            Some(i) if !i.is_empty() => {
                Ok(Identity { kind: kind.to_string(), instance: Some(i) })
            }
            _ => Err(SelectorError::InstanceRequired(kind.to_string())),
        };
    }
    if !SINGLE_INSTANCE_KINDS.contains(&kind) {
        return Err(SelectorError::UnknownKind(value.to_string()));
    }
    if instance.is_some() {
        return Err(SelectorError::InstanceNotAccepted(kind.to_string()));
    }
    Ok(Identity { kind: kind.to_string(), instance: None })
}

/// Reads the one variable from the process environment.
pub fn from_env() -> Result<Identity, SelectorError> {
    resolve(std::env::var(ENV_VAR).ok().as_deref())
}
